Sync up BlockNoteEditor with OP Color Theme

akabiru opened the pull request:
From https://github.com/opf/openproject/pull/19581

[blocknote-theme-sync.webm](https://github.com/user-attachments/assets/0fbd9b6c-c262-445c-8a9b-8d4a93c01a59)

akabiru committed:
Sync up BlockNoteEditor with OP Color Theme

brunopagno commented on frontend/src/react/OpBlockNoteContainer.tsx:
```diff
@@ -60,6 +68,26 @@ export default function OpBlockNoteContainer({ inputField, inputText }: OpBlockN
     ];
   };
 
+  useEffect(() => {
+    const observer = new MutationObserver((mutations) => {
+      mutations.forEach((mutation) => {
+        if (mutation.type === 'attributes' &&
+            (mutation.attributeName === 'data-color-mode' ||
+             mutation.attributeName === 'data-light-theme' ||
+             mutation.attributeName === 'data-dark-theme')) {
+          setTheme(detectTheme());
+        }
+      });
+    });
+
+    observer.observe(document.body, {
+      attributes: true,
+      attributeFilter: ['data-color-mode', 'data-light-theme', 'data-dark-theme']
+    });
+
+    return () => observer.disconnect();
+  }, []);
+
```
If I understood correctly this allows for real time updating the editor theme based on config changes. This feels unnecessary to me as it's unlikely to affect any users, really. I'd do the check on page load only.

That said I might be overoptimizing or oversimplifying :sweat_smile:

akabiru replied:
Good point. In the current flow "on load" should suffice but when/if we implement automatic theme changes via OS preference (which I'm missing out on) then this would be useful- if the sync happens while the user is on blocknote.

brunopagno replied:
Just documenting here what we talked sync~

BlockNote default behaviour should be picking up the configuration from the operating system, so we probably don't need a manual check here for live changes in the future.

brunopagno commented on frontend/src/react/OpBlockNoteContainer.tsx:
```diff
@@ -47,8 +47,16 @@ const schema = BlockNoteSchema.create({
+
 export default function OpBlockNoteContainer({ inputField, inputText }: OpBlockNoteContainerProps) {
   const [isLoading, setIsLoading] = useState(true);
+  const [theme, setTheme] = useState<"light" | "dark">(detectTheme);
```
Would be nice to define a type here for the light/dark

```typescript
type OpTheme = "light"|"dark"
...
const [theme, setTheme] = useState<OpTheme>(detectTheme);
```

bear with me that I did not check this on an editor, so syntax might be wrong

akabiru replied:
Nice one 👍🏾  yeah, need to switch off my "hackathon" brain 😄

brunopagno reviewed (approved):
Works as advertised. I like how simple this is :clap:

Left comments, but since all works nicely I'm approving now.

akabiru committed:
Use explicit types for `OpTheme`

akabiru committed:
Remove theme observer

Until we have OS theme preference syncing across the board, observing the document body data-theme-*
is a bit heavy handed.

https://github.com/opf/openproject/pull/19627/files#r2218387825

akabiru commented:
🛠️ Squash Commits on Merge

akabiru commented:
https://community.openproject.org/wp/65265
